### PR TITLE
python3Packages.pytorch: migrate to cudaPackages

### PR DESCRIPTION
**Pull request by SomeoneSerge**
Use the redistributable cuda packages, instead of runfile-based
cudatoolkit

~~Build log: https://gist.github.com/SomeoneSerge/4e5b5e9d9ee82c410eebad60d0cdc8f7~~

~~All went awry, it used cudatoolkit through cudnn. Reworking~~

EDIT later on 2022-04-15: Now it actually builds with cuda-redist! Rebased on the cudnn PR with removed propagatedBuildInputs, adjusted, and pushed

CC @NixOS/cuda-maintainers 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pytorch/default.nix [hunk 1]
```diff
@@ -156,7 +175,7 @@ in buildPythonPackage rec {
 
   preConfigure = lib.optionalString cudaSupport ''
     export TORCH_CUDA_ARCH_LIST="${lib.strings.concatStringsSep ";" final_cudaArchList}"
-    export CC=${cudatoolkit.cc}/bin/gcc CXX=${cudatoolkit.cc}/bin/g++
+    export CC=${cc}/bin/gcc CXX=${cc}/bin/g++
```
Comment (samuela) [inline pkgs/development/python-modules/pytorch/default.nix:178]: What is the motivation for this change? IIUC it's important to use a version of GCC consistent with the CUDA version. How do we handle this in the redist world? Do we have a mechanism? If not, we should look into adding one.

File: pkgs/development/python-modules/pytorch/default.nix [hunk 2]
```diff
@@ -30,23 +30,42 @@
...
 
 # assert that everything needed for cuda is present and that the correct cuda versions are used
-assert !cudaSupport || (let majorIs = lib.versions.major cudatoolkit.version;
+assert cudaSupport -> (let majorIs = cudaPackages.cudaMajorVersion;
                         in majorIs == "9" || majorIs == "10" || majorIs == "11");
 
 # confirm that cudatoolkits are sync'd across dependencies
-assert !(MPISupport && cudaSupport) || mpi.cudatoolkit == cudatoolkit;
-assert !cudaSupport || magma.cudatoolkit == cudatoolkit;
+# TODO: verify it's OK to equality-compare cudaPackages (which are attrsets, scopes, or whatever)
```
Comment (samuela) [inline pkgs/development/python-modules/pytorch/default.nix:41]: Yeah, that's a really interesting question. This was a nice ability when dealing with just one derivation. I have no idea how this works on package sets. Maybe @FRidh knows?

File: pkgs/development/python-modules/pytorch/default.nix (hunk 1 above)
Comment (samuela) [reply pkgs/development/python-modules/pytorch/default.nix:178]: EDIT: just saw the code comment above. Yeah, this is a little awkward and keeps us tied to cudatoolkit if we need to always be pulling out `cudaPackages.cudatoolkit.cc`, but I agree that issue is out of scope for this particular PR.

**Review by samuela — COMMENTED** [review]
hell yeah! this looks dope. does it build and everything?

File: pkgs/development/python-modules/pytorch/default.nix (hunk 2 above)
Comment (SomeoneSerge) [reply pkgs/development/python-modules/pytorch/default.nix:41]: My assumption was that Nix would shallowly compare the expressions, sort of as symbols. That'd be somewhat restrictive but tolerable: it'd essentially mean that users shouldn't mix cuda/cudnn instances coming from different evaluations of nixpkgs, but should write an overlay and evaluate a single consistent nixpkgs world

File: pkgs/development/python-modules/pytorch/default.nix (hunk 1 above)
Comment (SomeoneSerge) [reply pkgs/development/python-modules/pytorch/default.nix:178]: Yeah, we need to move the "compatible GCCs` list into the redist expression instead, otherwise it looks kind of weird

Comment (SomeoneSerge) [reply pkgs/development/python-modules/pytorch/default.nix:178]: Actually, I'm really confused why, but `nix-tree` shows that `cudatoolkit` is still part of the closure o_0

Comment (SomeoneSerge) [reply pkgs/development/python-modules/pytorch/default.nix:178]: Oh well... `cudnn.propagatedBuildInputs` 🤦🏻

**Comment by FRidh** [comment]
cudnn using libcublas instead of cudatoolkit if possible https://github.com/NixOS/nixpkgs/pull/168755

**Comment by SomeoneSerge** [comment]
Hooray!

**Comment by SomeoneSerge** [comment]
@FRidh same point arises: this assumes redist packages are present unconditionally, i.e. it only works for cuda>=11. I think we don't have to support cuda10 in downstream derivations

File: pkgs/development/python-modules/pytorch/default.nix [hunk 3]
```diff
@@ -30,23 +30,58 @@
...
+      cuda_cccl # <thrust/*>
+      cuda_nvprof # <cuda_profiler_api.h>
+      cuda_nvrtc
+      cuda_cudart
+      libcublas
+      libcufft
+      libcusolver
+      libcusparse
+      libcurand
+      nccl.dev
```
Comment (FRidh) [inline pkgs/development/python-modules/pytorch/default.nix:77]: it is preferable to use `lib.getDev` and `lib.getOut` over directly referring to outputs because this will keep functioning if outputs are being removed.

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 168745` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python310Packages.pytorchWithCuda</li>
    <li>python39Packages.pytorchWithCuda</li>
  </ul>
</details>

File: pkgs/development/libraries/science/math/magma/default.nix [hunk 4]
```diff
@@ -72,5 +72,8 @@ in stdenv.mkDerivation {
     maintainers = with maintainers; [ tbenst ];
   };
 
-  passthru.cudatoolkit = cudatoolkit;
+  passthru = {
+    # TODO: leave just cudaPackages
```
Comment (samuela) [inline pkgs/development/libraries/science/math/magma/default.nix:76]: What's the holdup on this change? Is this something you intend to do before merging or is this for the future?

File: pkgs/development/python-modules/pytorch/default.nix [hunk 5]
```diff
@@ -30,23 +30,58 @@
...
 
 # assert that everything needed for cuda is present and that the correct cuda versions are used
-assert !cudaSupport || (let majorIs = lib.versions.major cudatoolkit.version;
+assert cudaSupport -> (let majorIs = cudaPackages.cudaMajorVersion;
                         in majorIs == "9" || majorIs == "10" || majorIs == "11");
 
 # confirm that cudatoolkits are sync'd across dependencies
-assert !(MPISupport && cudaSupport) || mpi.cudatoolkit == cudatoolkit;
-assert !cudaSupport || magma.cudatoolkit == cudatoolkit;
+# TODO: verify it's OK to equality-compare cudaPackages (which are attrsets, scopes, or whatever)
```
Comment (samuela) [inline pkgs/development/python-modules/pytorch/default.nix:41]: Where do we stand on this question? Has anyone been able to run a quick test?

**Review by samuela — COMMENTED** [review]
I just have a few questions about the status of some of the TODOs, but otherwise looks great!

File: pkgs/development/python-modules/pytorch/default.nix (hunk 5 above)
Comment (SomeoneSerge) [reply pkgs/development/python-modules/pytorch/default.nix:41]: With help from [#nix-lang:nixos.org](https://matrix.to/#/#nix-lang:nixos.org):

- `==` evaluates terms into values
- ...and proceeds with [eqValues](https://github.com/NixOS/nix/blob/c9a29d0d92529338026190374d0e7e886f486e47/src/libexpr/eval.cc#L1602-L1607)

The latter:

- will probably immediately return `true` in absence of overrides, because of `&v1 == &v2` special case
- otherwise recurse intro the attrset
  - for attributes-derivations this would cause evaluation of their `outPath`s
  - for functions (`overrideScope'`?) it would immediately return `false`

TLDR: I think it means that we'll essentially require pointer-equality. I can be wrong

**Comment by SomeoneSerge** [comment]
Some nitpicks:
- cudatoolkit _somehow_ is still part of the _build_ closure (but not runtime closure)
- I'm wondering if with cuda 11.6 I could actually separate libs, includes and nvcc. I had an impression that in 11.6 the behaviour of `FindCUDAToolkit.cmake` is somewhat different

But maybe we should leave both of these questions for future

File: pkgs/development/libraries/science/math/magma/default.nix (hunk 4 above)
Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/magma/default.nix:76]: - Wasn't sure if someone relies on this passthru
- I was actually considering whether we should passthru our symlinkJoin under this name? I thought about reusing it for torchvision and torchaudio. Again, we can change this later

**Comment by samuela** [comment]
> cudatoolkit somehow is still part of the build closure (but not runtime closure)

Huh that's very odd! I wonder if there's a way to debug the build time dependency graph?

**Comment by SomeoneSerge** [comment]
> Huh that's very odd! I wonder if there's a way to debug the build time dependency graph?

I wonder if it could be because of `cudatoolkit.cc` 😆  I mean, if strings have contexts, maybe passthru has too

**Comment by samuela** [comment]
Ahh, yeah that's prob it!

**Comment by samuela** [comment]
Perhaps we should expose something like `cudaPackages_X_Y.cc` or `cudaPackages_X_Y.default-cc`?

**Comment by SomeoneSerge** [comment]
Yes, but in a different PR

**Comment by SomeoneSerge** [comment]
> 2 packages built:

Since you have them cached, could you please [check](https://gist.github.com/SomeoneSerge/2b18573dba91eaa9bb4e9e70b914d655) the runtime once more? My cache is out of date and I'm suddenly dreading I may have tested the wrong build or whatever 🙃

**Comment by samuela** [comment]
Yup, it works!

**Comment by SomeoneSerge** [comment]
> Yup, it works!

Something to keep in mind for future: if pytorch says `RuntimeError: Unable to find a valid cuDNN algorithm to run convolution`, what it really means is "you probably have hanging around a shell with tensorflow or jax and they've pre-allocated all available video memory"

**Comment by SomeoneSerge** [comment]
> cudatoolkit.cc

Ah, much it's much easier: it's because of magma, and it's part of runtime closure too

**Comment by samuela** [comment]
> Something to keep in mind for future: if pytorch says RuntimeError: Unable to find a valid cuDNN algorithm to run convolution, what it really means is "you probably have hanging around a shell with tensorflow or jax and they've pre-allocated all available video memory"

I've had something similar happen with JAX: it says can't find cuDNN... when in reality it just can't initialize cuDNN bc something else is hogging all the GPU memory

**Comment by SomeoneSerge** [comment]
I'm not sure I'm ready for this to be merged:

- I want to ensure that everything that is meant to support GPU still supports it. There are e.g. some suspicious new lines in the build log that probably ought be investigated: https://gist.github.com/SomeoneSerge/3b5d7e2e3af5318f4ef6a380f3ef5c35#file-pytorch-redist-log-diff-L160
- I maybe should rework magma too, otherwise between the PRs our pytorch will be including both distributions of cuda in the closure, which is not nice...

**Comment by SomeoneSerge** [comment]
Now magma uses redist packages too, reducing the reported  closure size of pytorch from `12.75GiB` to `10.15GiB` (which is still too much, imo)
Kineto/cupti issue resolved just with adding libcupti to the closure

**Comment by SomeoneSerge** [comment]
...note that this change doesn't support building magma with cuda10 (pre-cuda11.4)

We could account for those as well, but I thought it's ok to ignore because `magma` is outside `cudaPackages` namespace

**Comment by SomeoneSerge** [comment]
Interesting that `magma` has `platforms = unix`, but neither cudatoolkit nor redist cuda packages are marked with darwin support

**Comment by samuela** [comment]
> Interesting that magma has platforms = unix, but neither cudatoolkit nor redist cuda packages are marked with darwin support

Although I think it may technically be possible to hook up NVIDIA hardware to a macOS system, I've never heard of anyone actually doing this... except maybe for kicks?

**Comment by SomeoneSerge** [comment]
RE: `magma.cudaPackages == cudaPackages`

I've actually just noticed a bug with the original expression.
Evaluating `passthru` attributes of `pytorchWithCuda` on `master`[^1] triggers evaluation of `cudatoolkit`, resulting in an error:

[^1]: master is at `97eb40e0f8a9ec4e7306645dfc446f3cb4b9eed4` at the moment

```bash
❯ nix eval --impure --expr '((builtins.getFlake github:NixOS/nixpkgs/master).legacyPackages.x86_64-linux.python3Packages.pytorch).cudaSupport'
false
❯ nix eval --impure --expr '((builtins.getFlake github:NixOS/nixpkgs/master).legacyPackages.x86_64-linux.python3Packages.pytorchWithCuda).cudaSupport'
error: Package ‘cudatoolkit-11.6.1’ ... has an unfree license (‘unfree’), refusing to evaluate
```

This happens because of the following line in `master`:
```nix
assert !cudaSupport || magma.cudatoolkit == cudatoolkit;
```

So, just to verify, in this PR we don't seem to be repeating that same error, although the whole thing kind of smells

```bash
❯ nix eval --impure --expr '((builtins.getFlake (toString ./.)).legacyPackages.x86_64-linux.python3Packages.pytorch).cudaSupport'
false
❯ nix eval --impure --expr '((builtins.getFlake (toString ./.)).legacyPackages.x86_64-linux.python3Packages.pytorchWithCuda).cudaSupport'
true
```

**Comment by SomeoneSerge** [comment]
I suspect we might still run into similar evaluation issues if we make an overlay resulting in lack of pointer-equality 🤔

**Comment by SomeoneSerge** [comment]
```nix
pytorch = callPackage ../development/python-modules/pytorch {
    # ...
    cudaPackages = pkgs.cudaPackages.overrideScope' (final: prev: { });
}
```

Then `pytorchWithCuda.cudaSupport` results in

```
error: assertion '(cudaSupport -> ((magma).cudaPackages == cudaPackages))' failed
```

...and `pytorch.cudaSupport` evaluates into `false`

**Comment by FRidh** [comment]
Might want to write a function that takes a list of sets and a list of packages and tests for every package whether it is equal in every set.

**Comment by SomeoneSerge** [comment]
Just one thing I don't like about this is that the user cannot disable those asserts downstream, but I can live with that

**Comment by SomeoneSerge** [comment]
@FRidh 

> Might want to write a function that takes a list of sets and a list of packages and tests for every package whether it is equal in every set.

1. I have a feeling that referential equality is somehow more appropriate? Except we're relying on nix's internals
2. If we compare the packages, we'd be at risk of getting that `refusing to evaluate` error again 🤔

**Comment by SomeoneSerge** [comment]
We definitely should write our own equality function just for isolation, regardless of whether it falls back to `==` or tests only packages. This is out of scope of the PR, however

File: pkgs/development/python-modules/pytorch/default.nix [hunk 6]
```diff
@@ -30,23 +30,60 @@
...
-assert !cudaSupport || (let majorIs = lib.versions.major cudatoolkit.version;
+assert cudaSupport -> (let majorIs = cudaPackages.cudaMajorVersion;
                         in majorIs == "9" || majorIs == "10" || majorIs == "11");
 
-# confirm that cudatoolkits are sync'd across dependencies
-assert !(MPISupport && cudaSupport) || mpi.cudatoolkit == cudatoolkit;
-assert !cudaSupport || magma.cudatoolkit == cudatoolkit;
+# We expect referential equality of all cudaPackages used to ensure consistency
+# You can make an overlay and pass the same cudaPackages to pytorch, mpi, and magma
+# TODO: `==` is an implementation detail; move comparison logic to cudaPackages
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/pytorch/default.nix:42]: ^

**Comment by ConnorBaker** [comment]
@SomeoneSerge if you don't mind, I'd like to take this up and continue where you left off.

**Comment by SomeoneSerge** [comment]
@ConnorBaker this would be just great, please do!
